app/ui/browser/model/index.js

```diff
@@ -29,6 +34,8 @@ export const State = Immutable.Record({
   // This is a list of Page objects (see below)
   pages: Immutable.List(),
 
+  pageIds: Immutable.Set(),
```

Comment:
Another option that might work (and which could possibly eliminate the need for creating a `pageIds` set) is to move all of the frequently changing values out of the page object. For example, having a separate list of `statusTexts` which can be looked up via `page.id`. Then the page objects themselves shouldn't change as much and would give us the same effect as this `pageIds` array.